dspy/predict/avatar/avatar.py

```diff
@@ -0,0 +1,157 @@
+import dspy
+
+from copy import deepcopy
+from pydantic.fields import FieldInfo
+
+from dspy.predict.avatar.signatures import Actor
+from dspy.predict.avatar.models import Action, Tool
+from dspy.signatures.signature import ensure_signature
+
+
+class Avatar(dspy.Module):
+    def __init__(
+        self,
+        signature,
+        tools,
+        max_iters=3,
+        verbose=False,
+    ):
+        self.signature = ensure_signature(signature)
+        self.input_fields = self.signature.input_fields
+        self.output_fields = self.signature.output_fields
+
+        self.finish_tool = Tool(
+            tool=None,
+            name="Finish",
+            desc="returns the final output and finishes the task",
+        )
+
+        self.tools = tools + [self.finish_tool]
+        self.actor_signature = Actor
+
+        for field in list(self.input_fields.keys())[::-1]:
+            self.actor_signature = self.actor_signature.append(
+                field,
+                self._get_field(self.input_fields[field]),
+                type_=self.input_fields[field].annotation,
+            )
+
+        self.verbose = verbose
+        self.max_iters = max_iters
+        self.actor = dspy.TypedPredictor(self.actor_signature)
+
+        self.actor_clone = deepcopy(self.actor)
+
+
+    def _get_field(self, field_info: FieldInfo):
+        match field_info.json_schema_extra['__dspy_field_type']:
+            case 'input':
+                return dspy.InputField(
+                    prefix=field_info.json_schema_extra['prefix'],
+                    desc=field_info.json_schema_extra['desc'],
+                    format=field_info.json_schema_extra['format'] if 'format' in field_info.json_schema_extra else None,
+                )
+            case 'output':
+                return dspy.OutputField(
+                    prefix=field_info.json_schema_extra['prefix'],
+                    desc=field_info.json_schema_extra['desc'],
+                    format=field_info.json_schema_extra['format'] if 'format' in field_info.json_schema_extra else None,
+                )
+            case _:
+                raise ValueError(f"Unknown field type: {field_info.json_schema_extra['__dspy_field_type']}") 
+
+
+    def _update_signature(self, idx: int, omit_action: bool = False):
+        self.actor.signature = self.actor.signature.with_updated_fields(
+            f"action_{idx}", 
+            Action, 
+            __dspy_field_type="input"
+        )
+
+        self.actor.signature = self.actor.signature.append(
+            f"result_{idx}",
+            dspy.InputField(
+                prefix=f"Result {idx}:",
+                desc=f"{idx}th result",
```

Comment:
ah sometimes this might run into weird things like '1th', '2th', etc. maybe the description is not needed since the prefix is there?

Reply:
Hmmm can probably omit this out

Reply:
Hmmm this might be causing issue, removing this makes it trigger all the tools. Need the order to be incorporated in some form in desc ig.